04_Scheduling/03_Tasks_And_Threads.md

```diff
@@ -0,0 +1,132 @@
+* A reference to it's own virtual addressing space, that is a pointer to a PML4 table (this only if we are going to implement an os where every task has it's own address space)
+
+### Context Switching
+
+* Usually context switch is done during an IRQ (when it depends on design decision. 
+
+One way to achieve context switching, is to clone the current stack frame for the IRQ being served and update some of it's values:
+
+* rip - this will contain the pointer to the thread function
+* rflgas - will containg 0x202
+* rsp - is the stack pointer, since the pointer is growing backward 
+* // to add other registers to initialize
+
+The rip will contain the function that is associated with the thread, sp contains the newly allocated stack for it. 
+
+### Exiting the thread
+
+After the thread finish its execution, we need a way to make it exit gracefully (otherwise it start to run into garbage most likely...) so there are two possible scenarios:
+
+* The programmer has called a thread_exit function so in this case we are fine
+* The programmer didn't called the function and the thread has terminated finished the execution of the function, at this point if it will not be stopped it will run into garbage. 
+
+To achieve that we need to have an "exit" function to be executed after the function in the thread terminates it's execution. But how to do it? We need to check on the X86_64 Abi calling convention (or what architecture applies to you). As usual there are multiple ways to achieve that, the easiest one is to create a wrapper function that takes two parameters
+
+* the first is the function we want to execute
+* the second is the variable containing the arguments (we will go back to this later) 
+
+This function will call the wrapper function passing the argument as it is and after that will call our exit function. So the code will look similar to the following;
+
+```c
+void thread_execution_wrapper( void (*function)(void *), void *arg) {
+    function(arg);
+    _thread_exit();
+}
+```
+
+What should the exit function do? Again this depends on the design choice, and as usual there are multiple paths, it depends if we want to exit the thread as soon as it calls the the exit function, or let the scheduler do that, just updating its status to the one correspinding to a terminated task (let's call it DEAD status). 
+
+* If the choice is to delete the thread as soon as it exit terminate the function what will happen then is that the thread is placed into a DEAD state, then the function will take care of removing the task from the scheduler queue, freeing all resources (stack, execution frame, page tables, etc.) that are allocated to it, and after that it will free the memory allocated to the task itself. Remember that when a task call the exit function is still the one being executed
```

Comment:
I see what you're hinting at here, maybe its worth overtly saying that the danger here (as I found out haha) is that if you free the stack of the current thread, and your scheduler does not switch stacks, you will free the stack the cpu is currently using! You must be super careful with this approach, I think it might even be worth suggesting using the second option (what you use in dreamos64).